crates/uv-resolver/src/lock/mod.rs

```diff
@@ -207,12 +226,60 @@ impl Lock {
     }
 
     /// Remove wheels that can't be selected for installation due to environment markers.
-    fn remove_unreachable_wheels(requires_python: &RequiresPython, package: &mut Package) {
-        // Remove wheels that don't match `requires-python` and can't be selected for
-        // installation.
-        package
+    ///
+    /// For example, a package included under `sys_platform == 'win32'` does not need Linux
+    /// wheels.
+    fn remove_unreachable_wheels(
+        graph: &ResolutionGraph,
+        requires_python: &RequiresPython,
+        node_index: NodeIndex,
+        locked_dist: &mut Package,
+    ) {
+        // Remove wheels that don't match `requires-python` and can't be selected for installation.
+        locked_dist
             .wheels
             .retain(|wheel| requires_python.matches_wheel_tag(&wheel.filename));
+
+        // Filter by platform tags.
+
+        // See https://github.com/pypi/warehouse/blob/ccff64920db7965078cf1fdb50f028e640328887/warehouse/forklift/legacy.py#L100-L169
+        // for a list of relevant platforms.
+        let linux_tags = [
+            "manylinux1_",
+            "manylinux2010_",
+            "manylinux2014_",
+            "musllinux_",
+            "manylinux_",
+        ];
+        let windows_tags = ["win32", "win_arm64", "win_amd64", "win_ia64"];
+
+        locked_dist.wheels.retain(|wheel| {
+            // Naively, we'd check whether `platform_system == 'Linux'` is disjoint, or
+            // `os_name == 'posix'` is disjoint, or `sys_platform == 'linux'` is disjoint (each on its
+            // own sufficient to exclude linux wheels), but due to
+            // `(A ∩ (B ∩ C) = ∅) => ((A ∩ B = ∅) or (A ∩ C = ∅))`
+            // a single disjointness check with the intersection is sufficient, so we have one
+            // constant per platform.
+            let reachability_markers = &graph.reachability[&node_index];
```

Comment:
How does this work? Is this the full markers along all edges to the node?

Reply:
Don't you need `propagate_markers` here?

Reply:
> How does this work? Is this the full markers along all paths to the node?

Yes

> Don't you need propagate_markers here?

I didn't want to modify the graph and we need to consider the starting environments too, so it's separate.

Reply:
I don’t understand how this is correct without using marker propagation though.

Reply:
Wdym by marker propagation? `graph.reachability` is computed by `ResolutionGraph::reachability`, which determines for package the set of markers it can be reached by across the whole graph. It's similar to `propagate_markers` except that we include the starting fork environments and we don't modify the graph itself so we can still use the edge weights in `Lock::from_resolution_graph`.

I'll check if we can replace `propagate_markers` with using `reachability` or if `tool.uv.environments` causes problems with that.